File: desktop/macos/Desktop/Sources/Bluetooth/Session/DeviceSessionCoordinator.swift
import Foundation

enum DeviceSessionPhase: Equatable, Sendable {
  case idle
  case connecting
  case ready
  case disconnecting
  case waitingToReconnect(attempt: Int)

  var isConnecting: Bool {
    if case .connecting = self { return true }
    return false
  }

  var isReady: Bool {
    if case .ready = self { return true }
    return false
  }

  var allowsConnectionAttempt: Bool {
    switch self {
    case .idle, .waitingToReconnect:
      return true
    case .connecting, .ready, .disconnecting:
      return false
    }
  }
}

struct DeviceSessionSnapshot: Equatable {
  fileprivate(set) var generation: UInt64
  fileprivate(set) var phase: DeviceSessionPhase
  fileprivate(set) var pairedDevice: BtDevice?
  fileprivate(set) var connectedDevice: BtDevice?
  fileprivate(set) var failureMessage: String?

  static func initial(pairedDevice: BtDevice?) -> DeviceSessionSnapshot {
    DeviceSessionSnapshot(
      generation: 0,
      phase: .idle,
      pairedDevice: pairedDevice,
      connectedDevice: nil,
      failureMessage: nil
    )
  }
}

struct DeviceReconnectRequest {
  let device: BtDevice
  fileprivate let generation: UInt64
  fileprivate let attempt: Int
}

enum DeviceSessionCoordinatorError: LocalizedError, Equatable {
  case connectionAlreadyActive
  case connectionUnavailable
  case superseded

  var errorDescription: String? {
    switch self {
    case .connectionAlreadyActive:
      return "A device connection is already active"
    case .connectionUnavailable:
      return "The device is not currently available over Bluetooth"
    case .superseded:
      return "The device connection was superseded"
    }
  }
}

@MainActor
protocol DeviceSessionScheduledAction: AnyObject {
  func cancel()
}

@MainActor
protocol DeviceSessionScheduling: AnyObject {
  func schedule(
    after delay: Duration,
    action: @escaping @MainActor () -> Void
  ) -> any DeviceSessionScheduledAction
}

@MainActor
final class DeviceSessionTaskScheduler: DeviceSessionScheduling {
  private final class ScheduledAction: DeviceSessionScheduledAction {
    let task: Task<Void, Never>

    init(task: Task<Void, Never>) {
      self.task = task
    }

    func cancel() {
      task.cancel()
    }
  }

  private let clock: any DeviceOperationClock

  init(clock: any DeviceOperationClock = ContinuousDeviceOperationClock()) {
    self.clock = clock
  }

  func schedule(
    after delay: Duration,
    action: @escaping @MainActor () -> Void
  ) -> any DeviceSessionScheduledAction {
    let task = Task { @MainActor [clock] in
      do {
        if delay > .zero {
          try await clock.sleep(for: delay)
        }
      } catch {
        return
      }
      guard !Task.isCancelled else { return }
      action()
    }
    return ScheduledAction(task: task)
  }
}

/// Canonical owner for the logical lifecycle of one paired Bluetooth device.
///
/// Every connect attempt receives a monotonically increasing generation. The
/// coordinator accepts callbacks only from the current connection object and
/// generation, which makes late CoreBluetooth callbacks harmless.
@MainActor
final class DeviceSessionCoordinator: DeviceConnectionDelegate {
  typealias ConnectionFactory = @MainActor (BtDevice, UInt64) -> DeviceConnection?

  private let connectionFactory: ConnectionFactory
  private let scheduler: any DeviceSessionScheduling
  private let reconnectDelay: Duration
  private let autoReconnectEnabled: Bool

  private(set) var snapshot: DeviceSessionSnapshot
  private(set) var activeConnection: DeviceConnection?

  private var reconnectAttempt = 0
  private var scheduledReconnect: (any DeviceSessionScheduledAction)?

  var onSnapshotChanged: ((DeviceSessionSnapshot) -> Void)?
  var onReconnectRequested: ((DeviceReconnectRequest) -> Void)?
  var onDiscoveryRequested: (() -> Void)?
  var onSessionEnded: (() -> Void)?
  var onFallDetected: ((AccelerometerData) -> Void)?

  init(
    pairedDevice: BtDevice?,
    connectionFactory: @escaping ConnectionFactory,
    scheduler: any DeviceSessionScheduling,
    reconnectDelay: Duration = .seconds(15),
    autoReconnectEnabled: Bool = true
  ) {
    self.snapshot = .initial(pairedDevice: pairedDevice)
    self.connectionFactory = connectionFactory
    self.scheduler = scheduler
    self.reconnectDelay = reconnectDelay
    self.autoReconnectEnabled = autoReconnectEnabled
  }

  func connect(to device: BtDevice) async throws -> DeviceConnection {
    try await connect(to: device, reconnectRequest: nil)
  }

  func reconnect(_ request: DeviceReconnectRequest) async throws -> DeviceConnection {
    try await connect(to: request.device, reconnectRequest: request)
  }

  private func connect(
    to device: BtDevice,
    reconnectRequest: DeviceReconnectRequest?
  ) async throws -> DeviceConnection {
    let isReconnectAttempt = reconnectRequest != nil
    if let reconnectRequest {
      guard snapshot.generation == reconnectRequest.generation,
        snapshot.phase == .waitingToReconnect(attempt: reconnectRequest.attempt),
        snapshot.pairedDevice?.id == reconnectRequest.device.id
      else {
        throw DeviceSessionCoordinatorError.superseded
      }
    } else {
      guard snapshot.phase.allowsConnectionAttempt else {
        throw DeviceSessionCoordinatorError.connectionAlreadyActive
      }
    }

    cancelScheduledReconnect()
    let generation = nextGeneration()
    snapshot.phase = .connecting
    snapshot.connectedDevice = nil
    snapshot.failureMessage = nil
    publishSnapshot()

    guard let connection = connectionFactory(device, generation) else {
      let error = DeviceSessionCoordinatorError.connectionUnavailable
      recordConnectionFailure(error.localizedDescription)
      if isReconnectAttempt {
        scheduleReconnectIfNeeded(after: reconnectDelay)
      }
      throw error
    }

    activeConnection = connection
    connection.delegate = self

    do {
      try await connection.connect()
    } catch {
      let wasCurrent = isCurrent(connection)
      if wasCurrent {
        activeConnection = nil
        recordConnectionFailure(error.localizedDescription)
        if isReconnectAttempt {
          scheduleReconnectIfNeeded(after: reconnectDelay)
        }
      }
      await connection.disconnect()
      guard wasCurrent else {
        throw DeviceSessionCoordinatorError.superseded
      }
      throw error
    }

    guard isCurrent(connection), snapshot.phase.isConnecting else {
      await connection.disconnect()
      throw DeviceSessionCoordinatorError.superseded
    }

    reconnectAttempt = 0
    snapshot.phase = .ready
    snapshot.pairedDevice = connection.device
    snapshot.connectedDevice = connection.device
    snapshot.failureMessage = nil
    publishSnapshot()
    return connection
  }

  func disconnect(reconnectAfter delay: Duration? = .zero) async {
    cancelScheduledReconnect()
    guard let connection = activeConnection else {
      if let delay {
        scheduleReconnectIfNeeded(after: delay)
      }
      return
    }

    let teardownGeneration = nextGeneration()
    snapshot.phase = .disconnecting
    snapshot.connectedDevice = nil
    snapshot.failureMessage = nil
    publishSnapshot()

    activeConnection = nil
    await connection.disconnect()

    guard snapshot.generation == teardownGeneration,
      snapshot.phase == .disconnecting,
      activeConnection == nil
    else {
      return
    }
    snapshot.phase = .idle
    publishSnapshot()
    onSessionEnded?()

    if let delay {
      scheduleReconnectIfNeeded(after: delay)
    }
  }

  func unpair() async {
    cancelScheduledReconnect()
    let teardownGeneration = nextGeneration()
    let connection = activeConnection
    activeConnection = nil

    snapshot.phase = connection == nil ? .idle : .disconnecting
    snapshot.connectedDevice = nil
    snapshot.pairedDevice = nil
    snapshot.failureMessage = nil
    publishSnapshot()

    if let connection {
      await connection.unpair()

      guard snapshot.generation == teardownGeneration,
        snapshot.phase == .disconnecting,
        activeConnection == nil,
        snapshot.pairedDevice == nil
      else {
        return
      }
      onSessionEnded?()
    }

    snapshot.phase = .idle
    publishSnapshot()
  }

  func startReconnecting() {
    scheduleReconnectIfNeeded(after: .zero)
  }

  func stopReconnecting() {
    cancelScheduledReconnect()
    if case .waitingToReconnect = snapshot.phase {
      snapshot.phase = .idle
      publishSnapshot()
    }
  }

  func isReady(generation: UInt64) -> Bool {
    snapshot.generation == generation && snapshot.phase.isReady
  }

  func deviceConnection(
    _ connection: DeviceConnection,
    didDisconnectUnexpectedly device: BtDevice
  ) {
    guard isCurrent(connection) else { return }

    activeConnection = nil
    _ = nextGeneration()
    snapshot.phase = .idle
    snapshot.connectedDevice = nil
    snapshot.failureMessage = "Device disconnected unexpectedly"
    publishSnapshot()
    onSessionEnded?()
    scheduleReconnectIfNeeded(after: .zero)
  }

  func deviceConnection(
    _ connection: DeviceConnection,
    didDetectFall data: AccelerometerData
  ) {
    guard isCurrent(connection) else { return }
    onFallDetected?(data)
  }

  private func isCurrent(_ connection: DeviceConnection) -> Bool {
    activeConnection === connection
      && connection.sessionGeneration == snapshot.generation
  }

  @discardableResult
  private func nextGeneration() -> UInt64 {
    snapshot.generation &+= 1
    return snapshot.generation
  }

  private func recordConnectionFailure(_ message: String) {
    snapshot.phase = .idle
    snapshot.connectedDevice = nil
    snapshot.failureMessage = message
    publishSnapshot()
  }

  private func scheduleReconnectIfNeeded(after delay: Duration) {
    guard autoReconnectEnabled, let pairedDevice = snapshot.pairedDevice else { return }
    guard snapshot.phase.allowsConnectionAttempt else { return }

    cancelScheduledReconnect()
    reconnectAttempt += 1
    let expectedGeneration = snapshot.generation
    let expectedAttempt = reconnectAttempt
    snapshot.phase = .waitingToReconnect(attempt: expectedAttempt)
    publishSnapshot()

    // A zero-delay attempt is the direct reconnect path. Preserve the
    // manager's cached peripheral until that attempt has run; scanning
    // clears the cache. If direct reconnect cannot build a connection,
    // `connect` schedules the delayed path below, which starts discovery.
    if delay > .zero {
      onDiscoveryRequested?()
    }

    scheduledReconnect = scheduler.schedule(after: delay) { [weak self] in
      guard let self else { return }
      guard self.snapshot.generation == expectedGeneration,
        self.snapshot.phase == .waitingToReconnect(attempt: expectedAttempt),
        self.snapshot.pairedDevice?.id == pairedDevice.id
      else {
        return
      }
      self.scheduledReconnect = nil
      self.onReconnectRequested?(
        DeviceReconnectRequest(
          device: pairedDevice,
          generation: expectedGeneration,
          attempt: expectedAttempt
        )
      )
    }
  }

  private func cancelScheduledReconnect() {
    scheduledReconnect?.cancel()
    scheduledReconnect = nil
  }

  private func publishSnapshot() {
    onSnapshotChanged?(snapshot)
  }
}

File: desktop/windows/src/renderer/src/device/session/deviceSessionCoordinator.ts
/**
 * Device session coordinator — Windows port of macOS
 * Session/DeviceSessionCoordinator.swift. Owns the phase machine for one
 * paired wearable: which connection attempt is current, when a reconnect is
 * allowed, and which callbacks belong to a live generation.
 *
 * Every transition bumps the generation, so a callback from an abandoned
 * attempt can be recognized and dropped instead of mutating a newer session.
 */

import type { BtDevice } from '../protocol/btDevice'
import type { DeviceConnection } from '../connections/deviceConnection'
import { RealDeviceOperationClock, type DeviceOperationClock } from './deviceOperationBroker'

export type SessionPhaseKind =
  | 'idle'
  | 'connecting'
  | 'ready'
  | 'disconnecting'
  | 'waitingToReconnect'

export interface SessionPhase {
  kind: SessionPhaseKind
  /** Present on waitingToReconnect: which retry is pending. */
  attempt?: number
}

export interface DeviceSessionSnapshot {
  generation: number
  phase: SessionPhase
  pairedDevice: BtDevice | null
  connectedDevice: BtDevice | null
  failureMessage: string | null
}

export type DeviceSessionErrorKind =
  | 'connectionAlreadyActive'
  | 'connectionUnavailable'
  | 'superseded'

export class DeviceSessionError extends Error {
  readonly kind: DeviceSessionErrorKind

  private constructor(kind: DeviceSessionErrorKind, message: string) {
    super(message)
    this.name = 'DeviceSessionError'
    this.kind = kind
  }

  static connectionAlreadyActive(): DeviceSessionError {
    return new DeviceSessionError(
      'connectionAlreadyActive',
      'A device connection is already active'
    )
  }

  static connectionUnavailable(): DeviceSessionError {
    return new DeviceSessionError(
      'connectionUnavailable',
      'The device is not currently available over Bluetooth'
    )
  }

  static superseded(): DeviceSessionError {
    return new DeviceSessionError('superseded', 'The device connection was superseded')
  }
}

export interface DeviceReconnectRequest {
  device: BtDevice
  generation: number
  attempt: number
}

export interface DeviceSessionCallbacks {
  /** Builds a connection for one attempt; null when the device is not
   *  currently reachable (for example not in the discovery cache). */
  connectionFactory: (device: BtDevice, generation: number) => DeviceConnection | null
  onSnapshot?: (snapshot: DeviceSessionSnapshot) => void
  onSessionEnded?: () => void
  onReconnectRequested?: (request: DeviceReconnectRequest) => void
  /** Asked before a delayed reconnect so a scan can refresh the cache. */
  onDiscoveryRequested?: () => void
}

/** Fixed, not exponential: the mac stack retries at this cadence forever. */
export const RECONNECT_DELAY_MS = 15_000

const allowsConnectionAttempt = (phase: SessionPhase): boolean =>
  phase.kind === 'idle' || phase.kind === 'waitingToReconnect'

export class DeviceSessionCoordinator {
  private generation = 0
  private phase: SessionPhase = { kind: 'idle' }
  private pairedDevice: BtDevice | null
  private connectedDevice: BtDevice | null = null
  private failureMessage: string | null = null
  private activeConnection: DeviceConnection | null = null
  private reconnectAttempt = 0
  private reconnectAbort: AbortController | null = null

  autoReconnectEnabled = true

  constructor(
    private readonly callbacks: DeviceSessionCallbacks,
    options: { pairedDevice?: BtDevice | null; clock?: DeviceOperationClock } = {}
  ) {
    this.pairedDevice = options.pairedDevice ?? null
    this.clock = options.clock ?? new RealDeviceOperationClock()
  }

  private readonly clock: DeviceOperationClock

  get snapshot(): DeviceSessionSnapshot {
    return {
      generation: this.generation,
      phase: { ...this.phase },
      pairedDevice: this.pairedDevice,
      connectedDevice: this.connectedDevice,
      failureMessage: this.failureMessage
    }
  }

  get connection(): DeviceConnection | null {
    return this.activeConnection
  }

  // --- connect ---------------------------------------------------------------

  /**
   * Opens a session. A reconnect request must still match the generation,
   * phase, attempt and device it was scheduled for; anything else means a
   * newer decision has already replaced it.
   */
  async connect(device: BtDevice, reconnect?: DeviceReconnectRequest): Promise<void> {
    if (reconnect !== undefined) {
      const matches =
        reconnect.generation === this.generation &&
        this.phase.kind === 'waitingToReconnect' &&
        this.phase.attempt === reconnect.attempt &&
        this.pairedDevice?.id === reconnect.device.id &&
        // A retry token is for ONE device: without this, a valid token could be
        // replayed to open a session against a different device.
        reconnect.device.id === device.id
      if (!matches) throw DeviceSessionError.superseded()
    } else if (!allowsConnectionAttempt(this.phase)) {
      throw DeviceSessionError.connectionAlreadyActive()
    }

    this.cancelScheduledReconnect()
    const generation = ++this.generation
    this.phase = { kind: 'connecting' }
    this.connectedDevice = null
    this.failureMessage = null
    this.publish()

    const connection = this.callbacks.connectionFactory(device, generation)
    if (connection === null) {
      this.recordConnectionFailure(
        generation,
        DeviceSessionError.connectionUnavailable().message,
        reconnect !== undefined
      )
      throw DeviceSessionError.connectionUnavailable()
    }

    this.activeConnection = connection
    try {
      await connection.connect()
    } catch (error) {
      if (this.generation !== generation) {
        await connection.disconnect()
        throw DeviceSessionError.superseded()
      }
      this.activeConnection = null
      this.recordConnectionFailure(
        generation,
        error instanceof Error ? error.message : String(error),
        reconnect !== undefined
      )
      await connection.disconnect()
      throw error
    }

    if (this.generation !== generation || this.phase.kind !== 'connecting') {
      await connection.disconnect()
      throw DeviceSessionError.superseded()
    }

    this.reconnectAttempt = 0
    this.phase = { kind: 'ready' }
    this.pairedDevice = connection.device
    this.connectedDevice = connection.device
    this.publish()
  }

  private recordConnectionFailure(
    generation: number,
    message: string,
    wasReconnectAttempt: boolean
  ): void {
    if (this.generation !== generation) return
    this.phase = { kind: 'idle' }
    this.failureMessage = message
    this.publish()
    if (wasReconnectAttempt) {
      // The immediate retry already failed, so fall back to the delayed path,
      // which rescans before trying again.
      this.scheduleReconnectIfNeeded(RECONNECT_DELAY_MS)
    }
  }

  // --- disconnect / unpair ----------------------------------------------------

  /**
   * Tears the session down. reconnectAfterMs defaults to 0, which means
   * "reconnect immediately without scanning"; null means stay disconnected.
   */
  async disconnect(reconnectAfterMs: number | null = 0): Promise<void> {
    this.cancelScheduledReconnect()
    const connection = this.activeConnection
    if (connection === null) {
      if (reconnectAfterMs === null) {
        this.phase = { kind: 'idle' }
        this.publish()
      } else {
        this.scheduleReconnectIfNeeded(reconnectAfterMs)
      }
      return
    }

    const generation = ++this.generation
    this.activeConnection = null
    this.phase = { kind: 'disconnecting' }
    this.connectedDevice = null
    this.publish()

    await connection.disconnect()

    if (this.generation === generation) {
      this.phase = { kind: 'idle' }
      this.publish()
      this.callbacks.onSessionEnded?.()
    }
    if (reconnectAfterMs !== null) this.scheduleReconnectIfNeeded(reconnectAfterMs)
  }

  async unpair(): Promise<void> {
    this.cancelScheduledReconnect()
    const connection = this.activeConnection
    const generation = ++this.generation
    this.activeConnection = null
    this.connectedDevice = null
    this.pairedDevice = null
    this.phase = { kind: connection === null ? 'idle' : 'disconnecting' }
    this.publish()

    if (connection !== null) {
      await connection.unpair()
      if (this.generation === generation) {
        this.callbacks.onSessionEnded?.()
        this.phase = { kind: 'idle' }
        this.publish()
      }
    }
  }

  /** Called by the connection delegate; ignored unless it is the live one. */
  handleUnexpectedDisconnect(connection: DeviceConnection): void {
    if (this.activeConnection !== connection) return
    this.activeConnection = null
    this.generation += 1
    this.phase = { kind: 'idle' }
    this.connectedDevice = null
    this.failureMessage = 'Device disconnected unexpectedly'
    this.publish()
    this.callbacks.onSessionEnded?.()
    // Immediate: the cached peripheral is still valid right after a drop.
    this.scheduleReconnectIfNeeded(0)
  }

  // --- reconnect scheduling ---------------------------------------------------

  startReconnecting(): void {
    this.scheduleReconnectIfNeeded(0)
  }

  stopReconnecting(): void {
    this.cancelScheduledReconnect()
    if (this.phase.kind === 'waitingToReconnect') {
      this.phase = { kind: 'idle' }
      this.publish()
    }
  }

  private scheduleReconnectIfNeeded(delayMs: number): void {
    if (!this.autoReconnectEnabled) return
    const device = this.pairedDevice
    if (device === null) return
    if (!allowsConnectionAttempt(this.phase)) return

    this.cancelScheduledReconnect()
    this.reconnectAttempt += 1
    const attempt = this.reconnectAttempt
    const generation = this.generation
    this.phase = { kind: 'waitingToReconnect', attempt }
    this.publish()

    if (delayMs > 0) {
      // A delayed retry rescans first. The zero-delay path deliberately does
      // not: scanning clears the cached peripheral the direct reconnect needs.
      this.callbacks.onDiscoveryRequested?.()
    }

    const abort = new AbortController()
    this.reconnectAbort = abort
    void (async () => {
      if (delayMs > 0) {
        const outcome = await this.clock.sleep(delayMs, abort.signal)
        if (outcome !== 'elapsed') return
      }
      if (abort.signal.aborted) return
      const stillValid =
        this.generation === generation &&
        this.phase.kind === 'waitingToReconnect' &&
        this.phase.attempt === attempt &&
        this.pairedDevice?.id === device.id
      if (!stillValid) return
      this.callbacks.onReconnectRequested?.({ device, generation, attempt })
    })()
  }

  private cancelScheduledReconnect(): void {
    this.reconnectAbort?.abort()
    this.reconnectAbort = null
  }

  setPairedDevice(device: BtDevice | null): void {
    this.pairedDevice = device
    this.publish()
  }

  private publish(): void {
    this.callbacks.onSnapshot?.(this.snapshot)
  }
}
